File: src/src_platform_uwp/fmg/uwp/draw/img/win2d/MosaicsImg.cs
using System.Linq;
using System.ComponentModel;
using System.Collections.Generic;
using Windows.Graphics.Display;
using Windows.UI.Xaml;
using Microsoft.Graphics.Canvas;
using Microsoft.Graphics.Canvas.UI.Xaml;
using fmg.common;
using fmg.common.geom;
using fmg.core.img;
using fmg.core.mosaic.cells;
using fmg.uwp.utils;
using fmg.uwp.draw.mosaic;
using fmg.uwp.draw.mosaic.win2d;
using fmg.uwp.mosaic.win2d;

namespace fmg.uwp.draw.img.win2d {

   /// <summary> Representable <see cref="fmg.core.types.EMosaic"/> as image.
   /// <br/>
   /// Win2D impl
   /// </summary>
   public static class MosaicsImg {

      private const bool RandomCellBkColor = true;

      /// <summary> Representable <see cref="fmg.core.types.EMosaic"/> as image: common Win2D implementation part </summary>
      public abstract class AMosaicImgWin2D<TImage> : AMosaicsImg<TImage>
         where TImage : DependencyObject, ICanvasResourceCreator
      {
         protected readonly ICanvasResourceCreator _rc;
         private MosaicImgView _view;

         static AMosaicImgWin2D() {
            StaticRotateImgConsts.Init();
         }

         protected AMosaicImgWin2D(ICanvasResourceCreator resourceCreator) {
            _rc = resourceCreator;
         }

         protected class MosaicImgView : AMosaicViewWin2D {

            private AMosaicImgWin2D<TImage> _owner;
            public MosaicImgView(AMosaicImgWin2D<TImage> owner) {
               _owner = owner;
            }

            protected override PaintWin2DContext CreatePaintContext() {
               var cntxt = base.CreatePaintContext();
               cntxt.IconicMode = true;
               if (RandomCellBkColor)
                  cntxt.BkFill.Mode = 1 + ThreadLocalRandom.Current.Next(Mosaic.CellAttr.getMaxBackgroundFillModeValue());
               return cntxt;
            }

            public override SizeDouble Size {
               get {
                  var size = _owner.Size;
                  return new SizeDouble(size.Width, size.Height);
               }
            }

            public override void Invalidate(IEnumerable<BaseCell> modifiedCells = null) {
               Repaint(modifiedCells, null /* new RectDouble(_owner.Size.Width, _owner.Size.Height) */);
            }

            protected override void ChangeSizeImagesMineFlag() {
               // none...
            }

            protected override void Dispose(bool disposing) {
               if (Disposed)
                  return;

               base.Dispose(disposing);

               if (disposing)
                  _owner = null;
            }

         }

         protected MosaicImgView View {
            get {
               if (_view == null)
                  View = new MosaicImgView(this); // call this setter
               return _view;
            }
            set {
               if (_view != null)
                  _view.Dispose();
               _view = value;
               if (_view != null)
                  _view.Mosaic = Mosaic;
            }
         }

         protected PaintUwpContext<CanvasBitmap> PaintContext => View.PaintContext;

         protected override void OnPropertyChanged(PropertyChangedEventArgs ev) {
            //LoggerSimple.Put($">  OnPropertyChanged: {GetType().Name}.{Entity}: PropertyName={ev.PropertyName}");
            switch (ev.PropertyName) {
            case nameof(this.PaddingFull):
               PaintContext.Padding = PaddingFull;
               break;
            case nameof(this.BorderWidth):
               PaintContext.PenBorder.Width = BorderWidth;
               break;
            case nameof(this.BorderColor):
               var pb = PaintContext.PenBorder;
               pb.ColorLight = pb.ColorShadow = BorderColor;
               break;
            case nameof(this.BackgroundColor):
               PaintContext.BackgroundColor = BackgroundColor;
               break;
            }

            base.OnPropertyChanged(ev);

            if (RotateMode == ERotateMode.SomeCells) {
               switch (ev.PropertyName) {
               case nameof(this.Size):
                  ImageCache = null;
                  break;
               case nameof(this.RotatedElements):
               case nameof(this.BackgroundColor):
               case nameof(this.MosaicType):
               case nameof(this.SizeField):
                  _invalidateCache = true;
                  break;
               }
            }
         }

         protected override void DrawBody() {
            switch (RotateMode) {
            case ERotateMode.FullMatrix:
               DrawBodyFullMatrix();
               break;
            case ERotateMode.SomeCells:
               DrawBodySomeCells();
               break;
            }
         }

         #region PART ERotateMode.FullMatrix

         /// <summary> Return painted mosaic bitmap
         /// if (!OnlySyncDraw) {
         ///   Сама картинка возвращается сразу.
         ///   Но вот её отрисовка - в фоне.
         ///   Т.к. TImage есть DependencyObject, то его владелец может сам отслеживать отрисовку...
         /// }
         /// </summary>
         protected void DrawBodyFullMatrix() {
            View.Invalidate(Matrix);
         }

         #endregion

         #region PART ERotateMode.SomeCells

         private const bool UseCache = true;

         /// <summary> need redraw the static part of the cache </summary>
         private bool _invalidateCache = true;
         /// <summary>
         /// Cached static part of the picture.
         /// ! Recreated only when changing the original image size (minimizing CreateImage calls).
         /// </summary>
         private CanvasBitmap _imageCache;
         private CanvasBitmap ImageCache {
            set {
               if (!ReferenceEquals(_imageCache, value)) {
                  _imageCache?.Dispose();
                  _imageCache = value;
                  _invalidateCache = true;
               }
            }
            get {
               if (_imageCache == null) {
                  var dpi = DisplayInformation.GetForCurrentView().LogicalDpi;
                  ICanvasResourceCreator rc = Image;
                  ImageCache = new CanvasRenderTarget(rc, Size.Width, Size.Height, dpi);
               }
               if (_invalidateCache) {
                  _invalidateCache = false;
                  using (var ds = ((CanvasRenderTarget)_imageCache).CreateDrawingSession()) {
                     var save = View.Paintable;
                     View.Paintable = ds;
                     DrawCache();
                     View.Paintable = save; // restore
                  }
               }
               return _imageCache;
            }
         }

         /// <summary> copy cached image to original </summary>
         protected void CopyFromCache() {
            CanvasDrawingSession ds = View.Paintable;
            var rc = new Windows.Foundation.Rect(0, 0, Size.Width, Size.Height);
            ds.DrawImage(ImageCache, rc, rc, 1.0f, CanvasImageInterpolation.NearestNeighbor, CanvasComposite.Copy);
         }

         private void DrawCache() { DrawStaticPart(); }

         protected void DrawStaticPart() {
            View.PaintContext.IsUseBackgroundColor = true;

            IList<BaseCell> notRotated;
            if (!RotatedElements.Any()) {
               notRotated = Matrix;
            } else {
               var matrix = Matrix;
               var indexes = RotatedElements.Select(cntxt => cntxt.index).ToList();
               notRotated = new List<BaseCell>(matrix.Count - indexes.Count);
               int i = 0;
               foreach (BaseCell cell in matrix) {
                  if (!indexes.Contains(i))
                     notRotated.Add(cell);
                  ++i;
               }
            }
            View.Invalidate(notRotated);
         }

         protected void DrawRotatedPart() {
            if (!RotatedElements.Any())
               return;

            var pb = PaintContext.PenBorder;
            // save
            var borderWidth = BorderWidth;
            var borderColor = BorderColor;
            // modify
            pb.Width = 2 * borderWidth;
            pb.ColorLight = pb.ColorShadow = borderColor.Darker(0.5);

            View.PaintContext.IsUseBackgroundColor = false;
            var matrix = Matrix;
            var rotatedCells = new List<BaseCell>(RotatedElements.Count);
            foreach (RotatedCellContext cntxt in RotatedElements)
               rotatedCells.Add(matrix[cntxt.index]);
            View.Invalidate(rotatedCells);

            // restore
            pb.Width = borderWidth; //BorderWidth = borderWidth;
            pb.ColorLight = pb.ColorShadow = borderColor; //BorderColor = borderColor;
         }

         protected void DrawBodySomeCells() {
            if (UseCache)
               CopyFromCache();
            else
               DrawStaticPart();
            DrawRotatedPart();
         }

         #endregion

         protected override void Dispose(bool disposing) {
            if (Disposed)
               return;

            if (disposing) {
               View = null;
               ImageCache = null;
            }

            base.Dispose(disposing);
         }

      }

      /////////////////////////////////////////////////////////////////////////////////////////////////////
      //    custom implementations
      /////////////////////////////////////////////////////////////////////////////////////////////////////

      /// <summary> Representable <see cref="fmg.core.types.EMosaic"/> as image.
      /// <br/>
      /// CanvasBitmap impl
      /// </summary>
      public class CanvasBmp : AMosaicImgWin2D<CanvasBitmap> {

         public CanvasBmp(ICanvasResourceCreator resourceCreator)
            : base(resourceCreator)
         { }

         protected override CanvasBitmap CreateImage() {
            var dpi = DisplayInformation.GetForCurrentView().LogicalDpi;
            return new CanvasRenderTarget(_rc, Size.Width, Size.Height, dpi);
         }

         protected override void DrawBody() {
            using (var ds = ((CanvasRenderTarget)Image).CreateDrawingSession()) {
               View.PaintContext.IsUseBackgroundColor = true;
               View.Paintable = ds;
               base.DrawBody();
               View.Paintable = null;
            }
         }

      }

      /// <summary> Representable <see cref="fmg.core.types.EMosaic"/> as image.
      /// <br/>
      /// CanvasImageSource impl (XAML ImageSource compatible)
      /// </summary>
      public class CanvasImgSrc : AMosaicImgWin2D<CanvasImageSource> {

         public CanvasImgSrc(ICanvasResourceCreator resourceCreator /* = CanvasDevice.GetSharedDevice() */)
            : base(resourceCreator)
         { }

         protected override CanvasImageSource CreateImage() {
            var dpi = DisplayInformation.GetForCurrentView().LogicalDpi;
            return new CanvasImageSource(_rc, Size.Width, Size.Height, dpi);
         }

         protected override void DrawBody() {
            using (var ds = Image.CreateDrawingSession(BackgroundColor.ToWinColor())) {
               View.PaintContext.IsUseBackgroundColor = false;
               View.Paintable = ds;
               base.DrawBody();
               View.Paintable = null;
            }
         }

      }

   }

}

File: src/src_platform_swing/fmg/swing/draw/img/MosaicsImg.java
package fmg.swing.draw.img;

import java.util.ArrayList;
import java.util.Collection;
import java.util.List;
import java.util.Random;
import java.util.concurrent.ThreadLocalRandom;
import java.util.stream.Collectors;
import java.util.stream.Stream;

import fmg.common.Color;
import fmg.core.img.MosaicRotateTransformer;
import fmg.core.img.MosaicsAnimatedModel;
import fmg.core.img.MosaicsAnimatedModel.ERotateMode;
import fmg.core.img.MosaicsAnimatedModel.RotatedCellContext;
import fmg.core.mosaic.AMosaicController;
import fmg.core.mosaic.cells.BaseCell;
import fmg.core.mosaic.draw.MosaicDrawModel;
import fmg.core.types.EMosaic;
import fmg.data.view.draw.PenBorder;
import fmg.swing.mosaic.AMosaicViewSwing;

/**
 * Representable {@link fmg.core.types.EMosaic} as image
 * <br>
 * base SWING impl
 *
 * @param <TImage> SWING specific image: {@link java.awt.Image} or {@link javax.swing.Icon}
 */
public abstract class MosaicsImg<TImage>
                extends AMosaicViewSwing<TImage, Void, MosaicsAnimatedModel<Void>>
{

   private static final boolean RandomCellBkColor = true;

   protected MosaicsImg() {
      super(new MosaicsAnimatedModel<Void>());

      MosaicsAnimatedModel<Void> model = getModel();
      PenBorder pen = model.getPenBorder();
      pen.setColorLight(pen.getColorShadow());
      if (RandomCellBkColor)
         model.getBackgroundFill().setMode(1 + ThreadLocalRandom.current().nextInt(model.getCellAttr().getMaxBackgroundFillModeValue()));
   }

   @Override
   protected void onPropertyModelChanged(Object oldValue, Object newValue, String propertyName) {
      super.onPropertyModelChanged(oldValue, newValue, propertyName);

      MosaicsAnimatedModel<Void> model = getModel();
      /*
      switch (propertyName) {
      case MosaicGameModel.PROPERTY_CELL_ATTR:
         if (RandomCellBkColor)
            model.getBackgroundFill().setMode(1 + ThreadLocalRandom.current().nextInt(model.getCellAttr().getMaxBackgroundFillModeValue()));
         break;
      default:
         break;
      }
      */

      if (model.getRotateMode() == ERotateMode.someCells) {
         switch (propertyName) {
         case MosaicDrawModel.PROPERTY_SIZE:
            _imageCache = null;
            break;
         case MosaicsAnimatedModel.PROPERTY_ROTATED_ELEMENTS:
         case MosaicsAnimatedModel.PROPERTY_BACKGROUND_COLOR:
            _invalidateCache = true;
            break;
         }
      }
   }

   @Override
   protected void drawBody() {
      //super.drawBody(); // !override super implementtation
      switch (getModel().getRotateMode()) {
      case fullMatrix:
         drawBodyFullMatrix();
         break;
      case someCells:
         drawBodySomeCells();
         break;
      }
   }


   protected boolean _useBackgroundColor = true;
   protected boolean _drawOnCache = true;

   /** ///////////// ================= PART {@link ERotateMode#fullMatrix} ======================= ///////////// */

   private void drawBodyFullMatrix() {
      _drawOnCache = false;
      _useBackgroundColor = true;
      draw(getModel().getMatrix());
   }

   /** ///////////// ================= PART {@link ERotateMode#someCells} ======================= ///////////// */

   private static final boolean USE_CACHE = false; // true is not supported in SWING

   /** need redraw the static part of the cache */
   private boolean _invalidateCache = true;
   /**
    * Cached static part of the picture.
    * ! Recreated only when changing the original image size (minimizing CreateImage calls).
    **/
   private TImage _imageCache;
   protected TImage getImageCache() {
      if (_imageCache == null) {
         _imageCache = createImage();
         _invalidateCache = true;
      }
      if (_invalidateCache) {
         _invalidateCache = false;
         //Graphics save = get...();
         //getView().setPaintable(_imageCache as Graphics); // not implement in SWING :(
         drawCache();
         //set...(save); // restore
      }
      return _imageCache;
   }

   /** copy cached image to original */
   protected abstract void copyFromCache();

   protected void drawCache() { drawStaticPart(); }

   private void drawStaticPart() {
      MosaicsAnimatedModel<Void> model = getModel();

      List<BaseCell> notRotated;
      if (model.getRotatedElements().isEmpty()) {
         notRotated = model.getMatrix();
      } else {
         List<BaseCell> matrix = model.getMatrix();
         List<Integer> indexes = model.getRotatedElements().stream().map(cntxt -> cntxt.index).collect(Collectors.toList());
         notRotated = new ArrayList<>(matrix.size() - indexes.size());
         int i = 0;
         for (BaseCell cell : matrix) {
            if (!indexes.contains(i))
               notRotated.add(cell);
            ++i;
         }
      }
      draw(notRotated);
   }

   private void drawRotatedPart() {
      MosaicsAnimatedModel<Void> model = getModel();

      if (model.getRotatedElements().isEmpty())
         return;

      PenBorder pb = model.getPenBorder();
      // save
      int borderWidth = pb.getWidth();
      Color colorLight  = pb.getColorLight();
      Color colorShadow = pb.getColorShadow();
      // modify
      pb.setWidth(2 * borderWidth);
      pb.setColorLight(colorLight.darker(0.5));
      pb.setColorShadow(colorShadow.darker(0.5));

      List<BaseCell> matrix = model.getMatrix();
      List<BaseCell> rotatedCells = new ArrayList<>(model.getRotatedElements().size());
      for (RotatedCellContext cntxt : model.getRotatedElements())
         rotatedCells.add(matrix.get(cntxt.index));
      draw(rotatedCells);

      // restore
      pb.setWidth(borderWidth);
      pb.setColorLight(colorLight);
      pb.setColorShadow(colorShadow);
   }

   private void drawBodySomeCells() {
      _drawOnCache = USE_CACHE;
      _useBackgroundColor = true;
      if (USE_CACHE)
         copyFromCache();
      else
         drawStaticPart();

      _drawOnCache = false;
      _useBackgroundColor = false;
      drawRotatedPart();
   }

   /////////////////////////////////////////////////////////////////////////////////////////////////////
   //    custom implementations
   /////////////////////////////////////////////////////////////////////////////////////////////////////

   /** Moisac image view implementation over {@link javax.swing.Icon} */
   static class Icon extends MosaicsImg<javax.swing.Icon> {

      private IconSwing ico = new IconSwing(this);

      @Override
      protected javax.swing.Icon createImage() { return ico.create(); }

      @Override
      public void draw(Collection<BaseCell> modifiedCells) {
         draw(ico.getGraphics(), modifiedCells, null, _useBackgroundColor);
      }

      @Override
      protected void copyFromCache() {
       //drawImage(getImageCache(), 0, 0, getSize().width, getSize().height, null);
         throw new UnsupportedOperationException("not implemented...");
      }

      @Override
      public void close() {
         ico.close();
         super.close();
         ico = null;
      }

   }

   /** Mosaics image view implementation over {@link java.awt.Image} */
   static class Image extends MosaicsImg<java.awt.Image> {

      private ImageAwt img = new ImageAwt(this);

      @Override
      protected java.awt.Image createImage() { return img.create(); }

      @Override
      public void draw(Collection<BaseCell> modifiedCells) {
         img.draw(g -> draw(g, modifiedCells, null, _useBackgroundColor));
     }

      @Override
      protected void copyFromCache() {
       //drawImage(getImageCache(), 0, 0, getSize().width, getSize().height, null);
         throw new UnsupportedOperationException("not implemented...");
      }

   }

   /** Mosaics image controller implementation for {@link Icon} */
   public static class ControllerIcon extends AMosaicController<javax.swing.Icon, Void, MosaicsImg.Icon, MosaicsAnimatedModel<Void>> {
      public ControllerIcon() {
         super(new MosaicsImg.Icon());
         addModelTransformer(new MosaicRotateTransformer());
      }
   }

   /** Mosaics image controller implementation for {@link Image} */
   public static class ControllerImage extends AMosaicController<java.awt.Image, Void, MosaicsImg.Image, MosaicsAnimatedModel<Void>> {
      public ControllerImage() {
         super(new MosaicsImg.Image());
         addModelTransformer(new MosaicRotateTransformer());
      }
   }

   ////////////// TEST //////////////
   public static void main(String[] args) {
      Random rnd = ThreadLocalRandom.current();
      TestDrawing.testApp(() ->
//      // test single
//      Arrays.asList(new MosaicsImg.ControllerImage() { { setMosaicType(EMosaic.eMosaicSquare1); }})

         // test all
         Stream.of(EMosaic.values())

//               // variant 1
//               .map(e -> Stream.of(new MosaicsImg.ControllerIcon () { { setMosaicType(e); }},
//                                   new MosaicsImg.ControllerImage() { { setMosaicType(e); }}))
//               .flatMap(x -> x)

               // variant 2
               .map(e ->  rnd.nextBoolean()
                           ? new MosaicsImg.ControllerIcon () { { setMosaicType(e); }}
                           : new MosaicsImg.ControllerImage() { { setMosaicType(e); }})

               .collect(Collectors.toList())
      );
   }
   //////////////////////////////////

}
